qiskit/transpiler/preset_passmanagers/level3.py

```diff
@@ -340,8 +340,15 @@ def _contains_delay(property_set):
     ):
         pm3.append(_direction_check)
         pm3.append(_direction, condition=_direction_condition)
-    pm3.append(_reset)
-    pm3.append(_depth_check + _opt + _unroll, do_while=_opt_control)
+        pm3.append(_reset)
+        # For transpiling to a target we need to run GateDirection in the
+        # optimization loop to correct for incorrect directions that might be
+        # inserted by UnitarySynthesis which is direction aware but only via
+        # the coupling map which with a target doesn't give a full picture
+        pm3.append(_depth_check + _opt + _direction + _unroll, do_while=_opt_control)
```

Comment:
Since this is inside the optimization loop and only affects the case where a `target` is provided, can we omit this if possible? e.g.
```suggestion
        if target is not None:
            pm3.append(_depth_check + _opt + _direction + _unroll, do_while=_opt_control)
        else:
            pm3.append(_depth_check + _opt + _unroll, do_while=_opt_control)
```

Reply:
We can but the nesting gets a little weird because we'll have to `else`s at different levels with the same code under them. I'll do that now

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7227/commits/c6df7ef1d483951b5c98811a8170da998ee52fc1